# **Introducción a Pandas, MatPlotLib, Seaborn :**
#    **Dataframe: Demografia 2022**

*Preparado* por: Leandro González Támara <br>
Utilizando: <br>
![](https://www.python.org/static/community_logos/python-logo-generic.svg)

#***Pasos 1, 2, 3 y 4 para comenzar:*** <br>


In [1]:
# 2. Importar la libreria Pandas ;)
import pandas as pd

In [ ]:
# 3. Importar un archivo csv, si el delimitador es tabulación usar /t   si solo se quiere importar 2000 casos nrows=2000
# df=pd.read_csv('/content/drive/My Drive/Demografia2022/Demografia2022.csv', delimiter=';', encoding = 'unicode_escape' )

In [2]:
# 4. Importar un archivo de Excel (index_col=0 para escoger un columna predeterminada como index)
df=pd.read_excel("Demografia2022.xlsx", sheet_name="Demografia2022")

#***Pasos 5 a 11 para explorar el dataframe***

In [3]:
# 5. Ver las variables en el archivo
print(df.columns)

Index(['Country_name', 'TPopulation1Jan', 'TPopulation1July',
       'TPopulationMale1July', 'TPopulationFemale1July', 'PopDensity',
       'PopSexRatio', 'MedianAgePop', 'NatChange', 'NatChangeRT', 'PopChange',
       'PopGrowthRate', 'DoublingTime', 'Births', 'Births1519', 'CBR', 'TFR',
       'NRR', 'MAC', 'SRB', 'Deaths', 'DeathsMale', 'DeathsFemale', 'CDR',
       'LEx', 'LExMale', 'LExFemale', 'LE15', 'LE15Male', 'LE15Female', 'LE65',
       'LE65Male', 'LE65Female', 'LE80', 'LE80Male', 'LE80Female',
       'InfantDeaths', 'IMR', 'LBsurvivingAge1', 'Under5Deaths', 'Q5', 'Q0040',
       'Q0040Male', 'Q0040Female', 'Q0060', 'Q0060Male', 'Q0060Female',
       'Q1550', 'Q1550Male', 'Q1550Female', 'Q1560', 'Q1560Male',
       'Q1560Female', 'NetMigrations', 'CNMR', 'CCode', 'Region',
       'IncomeGroup', 'GDPpp2020', 'Country', 'Continent'],
      dtype='object')


In [4]:
# 6. Ver el tamaño del dataframe
df.shape
#df.shape[0]
#df.shape[1]

(237, 61)

In [5]:
# 7. Ver NULL
df.isnull().sum(axis = 0)

Country_name               0
TPopulation1Jan            0
TPopulation1July           0
TPopulationMale1July       0
TPopulationFemale1July     0
                          ..
Region                     0
IncomeGroup               23
GDPpp2020                 37
Country                    0
Continent                  0
Length: 61, dtype: int64

In [6]:
# 8. Ver el encabezado del dataframe
df.head()

Country_name  TPopulation1Jan  TPopulation1July  TPopulationMale1July  \
0     Afghanistan        40580.490         41128.771             20766.442   
1         Albania         2847.853          2842.321              1417.813   
2         Algeria        44543.592         44903.225             22862.237   
3  American Samoa           44.446            44.272                21.873   
4         Andorra           79.688            79.824                40.786   

   TPopulationFemale1July  PopDensity  PopSexRatio  MedianAgePop  NatChange  \
0               20362.329     63.3436     101.9846       16.8702   1162.402   
1                1424.507    103.7419      99.5301       37.5806     -3.068   
2               22040.987     18.8531     103.7260       28.0155    729.261   
3                  22.400    221.3625      97.6495       28.2539      0.454   
4                  39.038    169.8383     104.4777       42.9263      0.073   

   NatChangeRT  ...  Q1560Male  Q1560Female  NetMigrations    CNMR  CCode  \
0       28.237  ...   329.5630     203.0171        -65.846  -1.600    AFG   
1       -1.078  ...   118.0304      60.3632         -8.000  -2.810    ALB   
2       16.243  ...    91.6019      63.7084         -9.999  -0.223    DZA   
3       10.153  ...   173.9059     103.5929         -0.810 -18.130    ASM   
4        0.911  ...    46.7542      28.6944          0.200   2.505    AND   

                       Region          IncomeGroup     GDPpp2020  \
0                  South Asia           Low income   2078.479082   
1       Europe & Central Asia  Upper middle income  15646.041523   
2  Middle East & North Africa  Lower middle income  12037.481235   
3         East Asia & Pacific  Upper middle income           NaN   
4       Europe & Central Asia          High income  40897.330000   

          Country  Continent  
0     Afghanistan       Asia  
1         Albania     Europe  
2         Algeria     Africa  
3  American Samoa       Asia  
4         Andorra     Europe  

[5 rows x 61 columns]

In [7]:
# 9. Ver la finalización del dataframe
df.tail()

Country_name  TPopulation1Jan  TPopulation1July  \
232  Wallis and Futuna Islands           11.613            11.572   
233             Western Sahara          570.311           575.986   
234                Yemen, Rep.        33322.321         33696.614   
235                     Zambia        19745.738         20017.675   
236                   Zimbabwe        16152.942         16320.537   

     TPopulationMale1July  TPopulationFemale1July  PopDensity  PopSexRatio  \
232                 5.552                   6.021     82.6607      92.2183   
233               317.891                 258.094      2.1654     123.1685   
234             17023.203               16673.411     63.8230     102.0979   
235              9877.642               10140.033     26.9276      97.4123   
236              7705.602                8614.936     42.1883      89.4447   

     MedianAgePop  NatChange  NatChangeRT  ...  Q1560Male  Q1560Female  \
232       36.2699      0.058        4.974  ...    66.5377      62.1912   
233       31.2718      5.746       10.026  ...   166.2455     119.5720   
234       18.8385    778.579       23.095  ...   296.0545     171.0778   
235       17.0874    548.867       27.411  ...   359.5575     242.3784   
236       18.2356    345.195       21.143  ...   449.3632     325.3865   

     NetMigrations    CNMR  CCode                      Region  \
232         -0.136 -11.680    WLF         East Asia & Pacific   
233          5.600   9.770    ESH          Sub-Saharan Africa   
234        -29.998  -0.890    YEM  Middle East & North Africa   
235         -5.000  -0.250    ZMB          Sub-Saharan Africa   
236         -9.999  -0.612    ZWE          Sub-Saharan Africa   

             IncomeGroup    GDPpp2020                    Country  Continent  
232                  NaN          NaN  Wallis and Futuna Islands    Oceania  
233           Low income          NaN             Western Sahara     Africa  
234           Low income  3688.519409                      Yemen       Asia  
235           Low income  3623.854913                     Zambia     Africa  
236  Lower middle income  2444.478237                   Zimbabwe     Africa  

[5 rows x 61 columns]

In [8]:
# 10. Ver un casos del df
df.iloc[100:110]

Country_name  TPopulation1Jan  TPopulation1July  TPopulationMale1July  \
100  Isle of Man           84.434            84.519                41.859   
101       Israel         8970.733          9038.309              4508.098   
102        Italy        59119.400         59037.474             28788.875   
103      Jamaica         2828.991          2827.377              1402.379   
104        Japan       124278.310        123951.692             60239.861   
105       Jersey          110.273           110.778                54.682   
106       Jordan        11256.190         11285.868              5846.915   
107   Kazakhstan        19295.502         19397.998              9335.100   
108        Kenya        53499.879         54027.487             26777.549   
109     Kiribati          130.077           131.232                63.943   

     TPopulationFemale1July  PopDensity  PopSexRatio  MedianAgePop  NatChange  \
100                  42.660    147.7614      98.1235       45.4210     -0.158   
101                4530.212    417.6089      99.5119       29.0905    125.154   
102               30248.598    199.4678      95.1742       47.2653   -222.349   
103                1424.998    261.0690      98.4126       31.2957      7.773   
104               63711.831    328.9379      94.5505       48.7449   -753.228   
105                  56.096    954.9871      97.4785       39.9244      0.207   
106                5438.953    127.1017     107.5007       23.7162    200.547   
107               10062.898      7.1852      92.7675       29.5377    204.996   
108               27249.938     93.0341      98.2665       19.3816   1065.212   
109                  67.289    180.7596      95.0267       22.0339      2.699   

     NatChangeRT  ...  Q1560Male  Q1560Female  NetMigrations    CNMR  CCode  \
100       -1.877  ...    82.0652      52.2210          0.340   4.030    IMN   
101       13.855  ...    67.2107      34.4114          9.999   1.107    ISR   
102       -3.768  ...    56.2494      32.7781         58.496   0.991    ITA   
103        2.744  ...   211.6839     141.6682        -10.999  -3.883    JAM   
104       -6.079  ...    62.3703      33.4175         99.994   0.807    JPN   
105        1.870  ...    64.6759      40.3982          0.800   7.247    JEY   
106       17.659  ...   125.8263      79.8405       -141.192 -12.432    JOR   
107       10.570  ...   268.8189     125.8188          0.000   0.000    KAZ   
108       19.710  ...   407.9282     301.5383        -10.000  -0.185    KEN   
109       20.541  ...   209.4785     159.9510         -0.400  -3.043    KIR   

                         Region          IncomeGroup     GDPpp2020  \
100       Europe & Central Asia          High income           NaN   
101  Middle East & North Africa          High income  43721.623039   
102       Europe & Central Asia          High income  45935.996554   
103   Latin America & Caribbean  Upper middle income  10025.904795   
104         East Asia & Pacific          High income  42940.405252   
105       Europe & Central Asia                  NaN           NaN   
106  Middle East & North Africa  Upper middle income  10952.010609   
107       Europe & Central Asia  Upper middle income  28599.985878   
108          Sub-Saharan Africa  Lower middle income   5023.511272   
109         East Asia & Pacific  Lower middle income   2171.867943   

         Country  Continent  
100  Isle of Man     Europe  
101       Israel       Asia  
102        Italy     Europe  
103      Jamaica    America  
104        Japan       Asia  
105       Jersey     Europe  
106       Jordan       Asia  
107   Kazakhstan     Europe  
108        Kenya     Africa  
109     Kiribati    Oceania  

[10 rows x 61 columns]

In [9]:
# 11. Ordenar la base de datos por variables ascendente o descendente
#df.sort_values(by=['Births'],ascending=False).head() # Ordena todo el df
df[['Country','Continent','Births']].sort_values(by=['Births'],ascending=False).head(10) # Selecciona unas variables y ordena

Country Continent     Births
95                                India      Asia  23056.027
42                                China      Asia  10757.794
154                             Nigeria    Africa   8002.758
160                            Pakistan      Asia   6424.944
96                            Indonesia      Asia   4462.212
45   Congo, The Democratic Republic of     Africa   4133.987
66                             Ethiopia    Africa   3928.445
224                       United States   America   3726.867
16                           Bangladesh      Asia   2994.966
28                               Brazil   America   2723.266

### **Ejercicio 1:**<br>
Ordenar la base de datos de menor a mayor por 'IMR' solo seleccionar la variables Country, Region e IMR

In [10]:
# Solución:
df[['Country', 'Region', 'IMR']].sort_values(by=['IMR'], ascending=True)

Country                 Region      IMR
141                Montenegro  Europe & Central Asia   1.2045
92                  Hong Kong    East Asia & Pacific   1.3795
94                    Iceland  Europe & Central Asia   1.4157
193                  Slovenia  Europe & Central Asia   1.4177
70                    Finland  Europe & Central Asia   1.6280
..                        ...                    ...      ...
195                   Somalia     Sub-Saharan Africa  66.1650
39   Central African Republic     Sub-Saharan Africa  67.6984
118                   Lesotho     Sub-Saharan Africa  69.0136
189              Sierra Leone     Sub-Saharan Africa  69.9620
154                   Nigeria     Sub-Saharan Africa  71.8710

[237 rows x 3 columns]

#***Paso 12 para filtrar por variables o por casos***

In [11]:
# 12. Filtro por variables
df[['Country', 'GDPpp2020','IMR']].sort_values(by=['IMR'], ascending=True).tail(3)

Country    GDPpp2020      IMR
118       Lesotho  2682.235528  69.0136
189  Sierra Leone  1816.170271  69.9620
154       Nigeria  5459.158598  71.8710

In [12]:
# 12.1 para filtrar por texto en columna
df[df['Country'].str.contains('olo')]

Country_name  TPopulation1Jan  TPopulation1July  TPopulationMale1July  \
43          Colombia        51788.827         51874.024             25575.607   
194  Solomon Islands          716.218           724.272               369.900   

     TPopulationFemale1July  PopDensity  PopSexRatio  MedianAgePop  NatChange  \
43                26298.417     46.1641      97.2515       31.1710    338.317   
194                 354.373     25.1518     104.3814       19.3731     17.717   

     NatChangeRT  ...  Q1560Male  Q1560Female  NetMigrations   CNMR  CCode  \
43         6.512  ...   195.5519     101.5067       -167.924 -3.232    COL   
194       24.434  ...   182.7229     140.5665         -1.600 -2.206    SLB   

                        Region          IncomeGroup     GDPpp2020  \
43   Latin America & Caribbean  Upper middle income  16893.817185   
194        East Asia & Pacific  Lower middle income   2656.100603   

             Country  Continent  
43          Colombia    America  
194  Solomon Islands    Oceania  

[2 rows x 61 columns]

### **Ejercicio 2:**<br>
Filtrar la base de datos por 'Country' y por 'IMR' y decir el top 5 de países según tasa de mortalidad infantil

In [13]:
# Solución:
df[['Country', 'IMR']].sort_values(by=['IMR'], ascending=False).head(5)

Country      IMR
154                   Nigeria  71.8710
189              Sierra Leone  69.9620
118                   Lesotho  69.0136
39   Central African Republic  67.6984
195                   Somalia  66.1650

### **Ejercicio 3:**<br>
Decir el top 5 de países según tasa de mortalidad infantil en Latin America & Caribbean

In [14]:
# Solución
df[(df['Region'] == 'Latin America & Caribbean')].sort_values(by=['IMR'], ascending=False)[['Country', 'IMR']].head(5)

Country      IMR
90               Haiti  43.3752
24             Bolivia  24.9386
58  Dominican Republic  24.3508
89              Guyana  23.5595
85           Guatemala  19.8249

### **Ejercicio 4:**<br>
Filtrar los datos de Colombia

In [15]:
#Solución
df[df['Country'] == 'Colombia']['IMR']

43    10.886
Name: IMR, dtype: float64

#***Pasos 13 y 14 para hacer consultas básicas***

In [16]:
# 13.Tabular una variable cualitativa (frecuencias absolutas)
pd.value_counts(df['IncomeGroup'])

/tmp/ipykernel_108334/2798186124.py:2: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  pd.value_counts(df['IncomeGroup'])


IncomeGroup
High income            78
Upper middle income    54
Lower middle income    53
Low income             29
Name: count, dtype: int64

### **Ejercicio 5:**<br>
Tabular la variable 'Region':

In [17]:
# Solución:
df['Region'].value_counts()

Region
Europe & Central Asia              60
Sub-Saharan Africa                 52
Latin America & Caribbean          50
East Asia & Pacific                42
Middle East & North Africa         20
South Asia                          8
North America                       4
Middle East and Northern Africa     1
Name: count, dtype: int64

### **Ejercicio 6:**<br>
Tabular la variable Region en América:

In [35]:
# Solución:
df[(df['Region'] == 'North America') | (df['Region'] == 'Latin America & Caribbean')].value_counts()

Country_name              TPopulation1Jan  TPopulation1July  TPopulationMale1July  TPopulationFemale1July  PopDensity  PopSexRatio  MedianAgePop  NatChange  NatChangeRT  PopChange  PopGrowthRate  DoublingTime  Births    Births1519  CBR     TFR     NRR     MAC     SRB    Deaths   DeathsMale  DeathsFemale  CDR    LEx      LExMale  LExFemale  LE15     LE15Male  LE15Female  LE65     LE65Male  LE65Female  LE80     LE80Male  LE80Female  InfantDeaths  IMR      LBsurvivingAge1  Under5Deaths  Q5       Q0040     Q0040Male  Q0040Female  Q0060     Q0060Male  Q0060Female  Q1550     Q1550Male  Q1550Female  Q1560     Q1560Male  Q1560Female  NetMigrations  CNMR     CCode  Region                     IncomeGroup          GDPpp2020     Country                   Continent
Antigua and Barbuda       93.488           93.763            44.782                48.981                  213.0977    91.4273      35.0890       0.553      5.890        0.550      0.587          118.0830      1.124     0.101       11.98

### **Ejercicio 7:**<br>
Hacer una lista con todos los países de América que tienen un ingreso per cápita menor a 10.000 dólares

In [36]:
# Solución:
df[(df['Continent'] == 'America') & (df['GDPpp2020'] < 10000)][['Country','GDPpp2020']]

Country    GDPpp2020
20        Belize  6921.450076
24       Bolivia  9030.387681
51          Cuba  9099.670000
61   El Salvador  9668.727256
85     Guatemala  9769.238227
90         Haiti  3127.415571
91      Honduras  6252.569135
152    Nicaragua  6331.672499

In [37]:
# 14. Tabular una variable cualitativa (frecuencia absolura o relativa: True or False)
df['Continent'].value_counts(normalize=True).round(3)*100

Continent
Africa     26.2
Europe     25.3
America    22.8
Asia       17.3
Oceania     8.4
Name: proportion, dtype: float64

### **Ejercicio 8:**<br>
Determinar el porcentaje de paises por nivel de ingreso per cápita ('IncomeGroup')

In [38]:
df.sort_values(by=['IncomeGroup'])[['Country', 'IncomeGroup']]

Country  IncomeGroup
68                      Faroe Islands  High income
92                          Hong Kong  High income
169                       Puerto Rico  High income
170                             Qatar  High income
172                           Romania  High income
..                                ...          ...
181  Saint Vincent and the Grenadines          NaN
200                State of Palestine          NaN
211                           Tokelau          NaN
229                           Vatican          NaN
232         Wallis and Futuna Islands          NaN

[237 rows x 2 columns]

#***Pasos 15 a 19 para hacer estadísticas básicas***

In [39]:
# 14.1 Descriptivas de todas las variables cuantitativas
df.describe()

TPopulation1Jan  TPopulation1July  TPopulationMale1July  \
count     2.370000e+02      2.370000e+02            237.000000   
mean      3.350911e+04      3.365023e+04          16913.837992   
std       1.356770e+05      1.359462e+05          69553.380076   
min       5.080000e-01      5.100000e-01              0.246000   
25%       4.026470e+02      4.052720e+02            195.817000   
50%       5.439435e+03      5.489740e+03           2742.094000   
75%       2.179988e+04      2.212525e+04          11078.628000   
max       1.425925e+06      1.425887e+06         731180.498000   

       TPopulationFemale1July    PopDensity  PopSexRatio  MedianAgePop  \
count              237.000000    237.000000   237.000000    237.000000   
mean             16736.394713    467.641576   100.682988     31.218075   
std              66405.728034   2237.543396    16.388154      9.679166   
min                  0.265000      0.137600    81.683600     14.485800   
25%                214.167000     39.248900    95.847700     22.417800   
50%               2790.791000     96.983300    98.534800     31.566500   
75%              11186.771000    239.582100   101.309600     39.206700   
max             698243.300000  24475.838900   263.768400     57.531200   

          NatChange  NatChangeRT    PopChange  ...  Q0060Female       Q1550  \
count    237.000000   237.000000   237.000000  ...   237.000000  237.000000   
mean     282.251262     9.128911   282.251810  ...   137.807725   76.014505   
std      917.226484    10.721169  1028.234887  ...   100.986470   58.220800   
min     -980.855000   -15.366000 -7264.930000  ...    23.155200    9.091100   
25%        0.131000     1.181000     0.198000  ...    59.534800   31.548100   
50%       13.216000     7.674000    23.025000  ...   103.790100   61.094600   
75%      204.996000    16.139000   242.412000  ...   191.449500  101.846900   
max    10194.012000    37.195000  9706.711000  ...   484.501600  311.143400   

        Q1550Male  Q1550Female       Q1560   Q1560Male  Q1560Female  \
count  237.000000   237.000000  237.000000  237.000000   237.000000   
mean    93.145257    58.493290  153.198451  187.776845   117.852030   
std     66.569013    51.360045   93.278326  106.034052    83.280749   
min     10.192000     8.040400   23.648600   28.959300    18.463900   
25%     41.388300    21.269500   79.637800  104.067700    52.886700   
50%     77.855900    40.657100  134.241000  173.268100    90.978400   
75%    130.721800    74.789200  209.480800  256.368900   167.534400   
max    352.523900   264.994700  492.033900  561.016500   430.904700   

       NetMigrations        CNMR      GDPpp2020  
count     237.000000  237.000000     200.000000  
mean        0.000004    1.028595   23168.637570  
std       507.854509   17.528073   23917.155041  
min     -6673.580000 -155.070000     683.593750  
25%        -8.000000   -1.507000    5067.712346  
50%        -0.200000   -0.172000   14959.761950  
75%         1.000000    1.005000   34164.324185  
max      3366.387000  152.203000  134753.797686  

[8 rows x 55 columns]

In [40]:
# 14.2 Crear una variable a partir de una fórmula
df['PIB']=df['GDPpp2020']*df['TPopulation1July']

In [41]:
# Cambiar el formato a moneda de una variable, por ejemplo al PIB del mundo 2020
suma = df['PIB'].sum()*1000

# Ponerle formato de separador de miles a la suma
suma_formato_miles = '{:,.0f}'.format(suma)

# PIB con el formato de miles
print("PIB mundial 2020:", suma_formato_miles)

PIB mundial 2020: 147,897,689,278,685


In [42]:
# Cambiar a formato de moneda de las descriptivas
descriptivas=df.describe()  #applymap('{:,.0f}'.format)
formato = descriptivas.applymap('{:,.0f}'.format)
print(formato)

      TPopulation1Jan TPopulation1July TPopulationMale1July  \
count             237              237                  237   
mean           33,509           33,650               16,914   
std           135,677          135,946               69,553   
min                 1                1                    0   
25%               403              405                  196   
50%             5,439            5,490                2,742   
75%            21,800           22,125               11,079   
max         1,425,925        1,425,887              731,180   

      TPopulationFemale1July PopDensity PopSexRatio MedianAgePop NatChange  \
count                    237        237         237          237       237   
mean                  16,736        468         101           31       282   
std                   66,406      2,238          16           10       917   
min                        0          0          82           14      -981   
25%                      214         39   

/tmp/ipykernel_71314/586246951.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  formato = descriptivas.applymap('{:,.0f}'.format)


In [45]:
# 14.3 Descriptivas de variables cuantitativas con formato de miles
descriptivas = df[['TPopulation1July','PIB','GDPpp2020','IMR','CBR']].describe()
# Formatear la salida con el separador de miles, nota .3f da el número de decimales
formato = descriptivas.applymap('{:,.0f}'.format)
print(formato)

      TPopulation1July             PIB GDPpp2020  IMR  CBR
count              237             200       200  237  237
mean            33,650     739,488,446    23,169   17   18
std            135,946   2,745,408,927    23,917   17    9
min                  1          57,493       684    1    4
25%                405      21,400,472     5,068    5   10
50%              5,490      84,886,569    14,960   10   14
75%             22,125     436,404,094    34,164   24   23
max          1,425,887  27,574,143,101   134,754   72   45


/tmp/ipykernel_71314/3154763871.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  formato = descriptivas.applymap('{:,.0f}'.format)


In [46]:
# 14.4 Con la opción suma se puede calcular la población mundial al 1 de julio de 2022
poblacion_mundial=df['TPopulation1July'].sum()*1000
poblacion_mundial2 = '{:,.0f}'.format(poblacion_mundial)
print('La población mundial en 2022 era de',poblacion_mundial2, 'habitantes')

La población mundial en 2022 era de 7,975,105,152 habitantes


In [47]:
# 15. Descriptivas básicas por variable
df['TFR'].describe().round(1)

count    237.0
mean       2.4
std        1.2
min        0.8
25%        1.6
50%        2.0
75%        3.0
max        6.7
Name: TFR, dtype: float64

Hay 237 países con información de la tasa de fertilidad total.

---

El promedio de las tasas de fertilidad en el mundo es de aprox. 2.4 nacidos vivos por mujer y su desviación estándar es de aprox.1.2 nacidos vivos por mujer (la diferencia promedio de los países con respecto a la media global de 2.4 es 1.2 nacidos vivos).
Los valores mínimo y máximo de la tasa de fertilidad en el mundo son 0.8 y 6.7  nacidos vivos por mujer aprox., que corresponden a Hong Kong y Niger respectivamente.
El percentil 75 de la tasa de fertilidad total es aprox. 3 nacidos vivos por mujer, esto significa que el 75% de los países tienen una tasa de fertilidad total menor o igual a 3, o lo que es lo mismo, que el 25% de los países tienen una tasa de fertilidad superior a 3 nacidos vivos por mujer.

### **Ejercicio. Analizar las estadísticas descriptivas del ingreso per cápita en el mundo - año 2020**

In [53]:
# Solución:
(df['GDPpp2020']*1000).describe().round(1).apply('{:,.0f}'.format)

count            200
mean      23,168,638
std       23,917,155
min          683,594
25%        5,067,712
50%       14,959,762
75%       34,164,324
max      134,753,798
Name: GDPpp2020, dtype: object

Hay 200 países con información del ingreso per cápita en el mundo

---

El promedio de lo que le corresponde a cada persona en el mundo 


### **Ejercicio 9:**<br>
Encuentre e interprete las estadísticas básicas de la fertilidad en el mundo

In [ ]:
# Solución:


### **Ejercicio 10:**<br>
Encuentre e interprete las estadísticas básicas de la tasa de mortalidad 'CDR' en Africa Subsahariana

In [ ]:
# Solución:


In [ ]:
# 16. Descriptivas específicas por variable. Las opciones:.count(), nunique(), .min(), .max(), .sum(), .mean(), .median, .mode(), .std(), .var(), .quantile()
df['GDPpp2020'].quantile(.05).round(1)

In [ ]:
# 17. Resumen estadístico de variables seleccionoadas del df
round(df[['CDR','CBR','IMR']].describe(),1)

### **Ejercicio 11:**<br>
Encuentre el total de nacimientos de latinoamérica y el caribe

In [ ]:
# Solución


### **Ejercicio 11.1:**<br>
Entre China y Latinoamérica, ¿en dónde hay mayor población?

In [ ]:
# Solución


### **Ejercicio 12:**<br>
Encuentre el pais de Europe & Central Asia con mayor ingreso per cápita

In [ ]:
#Solución


### **Ejercicio 13:**<br>
Encuentre las estadísticas descriptivas de las variables CDR, IMR y PopGrowthRate en Latinoamérica y el Caribe

In [ ]:
# Solución


In [ ]:
# 18. Agrupaciones (Tablas dinámicas)
df.groupby(['Continent','Region'])['TFR'].aggregate(['count','mean','std']).head(12).round(1)

In [ ]:
# 19. Tablas de contingencia
pd.crosstab(df["Continent"], df["IncomeGroup"],normalize=False)
#pd.crosstab(df["Continent"], df["IncomeGroup"],normalize=True)
#pd.crosstab(df["Continent"], df["IncomeGroup"],normalize='columns')
#pd.crosstab(df["Continent"], df["IncomeGroup"],normalize='index')*100

### **Ejercicio 14:**<br>
Hacer una tabla con las descriptivas básicas de la tasa cruda de nacimientos por región en África

In [ ]:
# Solución:


### **Ejercicio 15:**<br>

Diga cuál es el promedio del ingreso per cápita en Asia y compárelo con el promedio mundial

In [ ]:
#Solución:


### **Ejercicio 16:**<br>

Diga la región de Africa donde hay mayor variación en la tasa de nacimientos y señale el país que más aporta a esta variación.

In [ ]:
#Solución


#**Pasos con variables. Transformar - Recodificar**

In [ ]:
# 19.1 Para categorizar una variable cuantitativa
df['Fertilidad'] = pd.cut(df['TFR'], bins=[0, 1.571600, 2.003400, 2.979900,7], labels=['Baja', 'Media', 'Alta','Muy alta'])
pd.value_counts(df['Fertilidad'])

In [ ]:
# 19.2 Para recodificar los valores de una variable cualitativa
df['Continent'] = df['Continent'].replace({'Europe': 'Europa', 'America': 'Americano'})
pd.value_counts(df['Continent'])

Africa       62
Europa       60
Americano    54
Asia         41
Oceania      20
Name: Continent, dtype: int64

In [ ]:
df = df.set_index('Name').reset_index()

#***Pasos Visualización de datos***

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# 20. Hacer una gráfico de barras de una variable cualitativa ([:n] escoge cuantas modalidades incluir)
ax = df['Region'].value_counts()[:22].plot(kind='barh', figsize=(10,5), title="Numero de paises por región", color='thistle')
ax.set_xlabel("frecuencia");

In [ ]:
# 21. Hacer una gráfico de barras de una variable cualitativa ([:n] escoge cuantas modalidades incluir) (use figsize=(15,12) para full screen)
ax = df['Continent'].value_counts()[:5].plot(kind='barh', figsize=(10,5), title="Número de países por Continente",color='deepskyblue')
ax.set_xlabel("Frecuencia")
ax.set_ylabel("Continente");

In [ ]:
# 22. Hacer una gráfico de barras de una variable cualitativa ([:n] escoge cuantas modalidades incluir)
ax = df['Region'].value_counts()[:22].plot(kind='barh', figsize=(10,5), title="Numero de paises por región", color='coral')
ax.set_xlabel("frecuencia")

In [ ]:
# 23. Otra manera de hacer unas barras
import matplotlib.pyplot as plt
df.groupby('Continent')['Country'].nunique().plot(kind='barh', figsize=(10,5), color='plum')
plt.show()

### **Ejercicio 17:**<br>
Construir un gráfico de barras con el número de paises por nivel de ingreso per cápita

In [ ]:
# Solución:


In [ ]:
# Solución:


### **Ejercicio 18:**<br>
Construir un gráfico de barras con el promedio de la fertilidad por Continente

In [ ]:
# Solución:


### **Ejercicio 19:**<br>
Construir un gráfico de barras con el la desviación estándar del ingreso per cápita por región

In [ ]:
# Solución:


In [ ]:
# 24. Un grafico de barras bivariado
import matplotlib.pyplot as plt
df.groupby(['Continent','IncomeGroup']).size().unstack().plot(kind='bar',stacked=False, figsize=(15, 5))
plt.show()


In [ ]:
# 25. Un histograma básico
fig, ax = plt.subplots(figsize=(15,5))
plt.hist(df['GDPpp2020'], bins=10, color='blue',density=False);

### **Ejercicio 20:**<br>
Construir e interpretar un histograma de la fertilidad

In [ ]:
# Solución:
fig, ax = plt.subplots(figsize=(15,5))
plt.hist(df['TFR'], bins=10, color='violet');

### **Ejercicio 21:**<br>
Construir e interpretar un histograma de la esperanza de vida

In [ ]:
# Solución:


In [ ]:
# 26. Dos histogramas superpuestos
import seaborn as sns
import matplotlib.pyplot as plt
fig, ax = plt.subplots(figsize=(15,5))
sns.histplot( df["GDPpp2020"][df['Continent']=='Africa'] ,bins=7, color="yellow", label="Africa",kde=False)
sns.histplot( df["GDPpp2020"][df['Continent']=='Europa'] ,bins=7, color="blue", label="Europe",kde=False)
ax.set_xlabel("fertilidad")
plt.legend();

### **Ejercicio 22:**<br>
Comparar con dos histogramas superpuestos la fertilidad de Africa y la de Europe

In [ ]:
# Solución:


In [ ]:
# 27. Gráfico de dispersión con matplotlib
import numpy as np
import matplotlib.pyplot as plt
df.plot(kind='scatter',x='GDPpp2020',y='TFR',color='red', s=205, alpha=.3, figsize=(15, 5))
plt.show()

In [ ]:
# 28. Gráfico de dispersión con seaborn (se le puede agregar size='variable')
import seaborn as sns
fig, ax = plt.subplots(figsize=(15,7))
ax = sns.scatterplot(x='GDPpp2020', y="TFR", hue='Continent', s=300, data=df)

### **Ejercicio 23:**<br>
Interpretar un gráfico de dispersión entre el ingreso per cápita y la esperanza de vida

In [ ]:
# Solución
import numpy as np
import matplotlib.pyplot as plt
df.plot(kind='scatter',x='GDPpp2020',y='LEx',color='green', s=205, alpha=.3, figsize=(15, 5))
plt.show()

### **Ejercicio 24:**<br>
Interpretar un gráfico de dispersión entre la ingreso per cápita y la tasa de mortalidad infantil

In [ ]:
# Solución
import numpy as np
import matplotlib.pyplot as plt
df.plot(kind='scatter',x='GDPpp2020',y='IMR',color='blue', s=205, alpha=.3, figsize=(15, 5))
plt.show()

### **Ejercicio 25:**<br>
Interpretar un gráfico de dispersión entre la ingreso per cápita y la tasa de crecimiento de la población y marcar los puntos con el Continente.

In [ ]:
# Solución:
import seaborn as sns
fig, ax = plt.subplots(figsize=(15,7))
ax = sns.scatterplot(x='GDPpp2020', y="PopGrowthRate", hue='Continent', s=200, data=df)

In [ ]:
# 30. Stripplot
fig, ax = plt.subplots(figsize=(15,5))
m = sns.stripplot(x="Region", y="GDPpp2020", data=df, palette="Set3", s=10, ax=ax)
plt.xticks(rotation=90);
plt.title('Ingreso per cápita por región');

### **Ejercicio 26:**<br>
Interpretar un stripplot del nivel de ingreso per cápita y la fertilidad

In [ ]:
# Solución:


In [ ]:
# 31. Mapas de calor de variables cuantitativa
# Opciones para kind = “scatter” | “reg” | “resid” | “kde” | “hex”
sns.jointplot(x=df["GDPpp2020"], y=df["TFR"], kind='kde', color="olive", height=6).fig.set_figwidth(15)


### **Ejercicio 27:**<br>
Construir un mapa de calor de la esperanza de vida contra la tasa de mortalidad infantil

In [ ]:
# Solución:


In [ ]:
# 32. Grafica de caja y bigotes
fig, ax = plt.subplots(figsize=(15,5))
sns.boxplot(x="Continent", y="LEx", data=df, ax=ax);
plt.xticks(rotation=90);

### **Ejercicio 28:**<br>
Construir un boxplot de la fertilidad por Continente

In [ ]:
#Solución:


### **Ejercicio 29:**<br>
Construir un boxplot de la fertilidad por Continente y por nivel de ingreso per cápita

In [ ]:
# Solución:


In [ ]:
# 33. Tablas dinámicas (probar interpretación con media, suma, desviación)
tablapivote1=df.pivot_table(index='Region',columns='IncomeGroup',values='TFR',aggfunc='mean')
tablapivote1

### **Ejercicio 30:**<br>
Construir una tabla dinámica del promedio de la fertilidad por región y nivel de ingreso per cápita.

In [ ]:
# Solución


In [ ]:
# 34. Mapa de calor
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(style="whitegrid")
fig, ax = plt.subplots(figsize=(15,5))
sns.heatmap(tablapivote1, cmap="coolwarm", ax=ax);
ax.set_title('Mapa de calor del promedio de la fertilidad');

### **Ejercicio 31:**<br>
Construir un mapa de calor del promedio de la esperanza de vida por región y nivel de ingreso per cápita

In [ ]:
# Solución


In [ ]:
# 35.Gráfico de dispersión múltiple
import seaborn as sns
sns.pairplot(df, hue='Continent', x_vars=['GDPpp2020','TFR'], y_vars=['GDPpp2020','TFR'], plot_kws={'s':210},height=5, palette='bright');

### **Ejercicio 32:**<br>
Construir un gráfico de dispersión múltiple con las variables: GDPpp2020, TFR, LEx, IMR

In [ ]:
# Solución


In [ ]:
# 36. Un boxplot simple
fig, ax = plt.subplots(figsize=(15,5))
sns.boxplot(x="Births", y="Continent", data=df, ax=ax);
plt.xticks(rotation=90);

In [ ]:
# 38. Crear un data frame de manera manual
datos1 = [['maria', 21, 49], ['pao', 26, 52], ['juli', 24, 55], ['eliana',40, 60]]
  # Crear el df
df1 = pd.DataFrame(datos1, columns = ['nombre', 'edad','peso'])
df1

### **Ejercicio 33:**<br>
Convierta esta tabla en un dataframe de pandas

| sexo   | talla | peso |
|--------|-------|------|
| MUJER  | 159   | 49   |
| HOMBRE | 164   | 62   |
| MUJER  | 172   | 65   |
| MUJER  | 167   | 52   |
| MUJER  | 164   | 51   |
| MUJER  | 161   | 67   |
| MUJER  | 168   | 48   |
| HOMBRE | 181   | 74   |
| HOMBRE | 183   | 74   |
| MUJER  | 158   | 50   |
| MUJER  | 156   | 65   |
| HOMBRE | 173   | 64   |
| MUJER  | 158   | 43   |
| HOMBRE | 178   | 74   |

In [ ]:
# Solución:
gimnasio=[['MUJER',159,49],['HOMBRE',164,62],['MUJER',172,65],['MUJER',167,52],['MUJER',164,51],['MUJER',161,67],['MUJER',168,48],['HOMBRE',181,74],['HOMBRE',183,74],['MUJER',158,50],['MUJER',156,65],['HOMBRE',173,64],['MUJER',158,43],['HOMBRE',178,74]]
gym = pd.DataFrame(gimnasio, columns = ['sexo', 'talla','peso'])
gym

#***Correlación***

In [ ]:
corr = df['GDPpp2020'].corr(df['IMR'])
corr

-0.5969102377646225

#***Distribución binomial***

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

# Set the parameters for the binomial distribution
n = 10  # number of trials
p = 0.2  # probability of success

# Generate the probability mass function of the binomial distribution
x = np.arange(0, n+1)
pmf = np.zeros(n+1)
for i in range(n+1):
    pmf[i] = np.math.comb(n, i) * p**i * (1-p)**(n-i)

# Plot the probability mass function of the binomial distribution
fig, ax = plt.subplots(figsize=(15,5))
plt.stem(x, pmf, use_line_collection=True)

# Set the title and axis labels
plt.title(f'Binomial Distribution, n={n}, p={p}')
plt.xlabel('Number of successes')
plt.ylabel('Probability')

# Show the plot
plt.show()


In [ ]:
# Función de distribución binomial acumulada
from scipy.stats import binom
n = 10  # ensayos
p = 0.2  # probabilidad de exito
k = 6  # valor hasta el cual se acumula la distribución
cdf = binom.cdf(k, n, p)  # Función de distribución hasta k
print(cdf)


In [ ]:
from scipy.stats import poisson
k= 4 # valor hasta el cual se acumula
media=4
poisson.cdf(k,media)

0.6288369351798734

#***Distribuciones normales***

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
# Establecer la media y la desviación de la distribución
mu = 25  # media
sigma = 2  # desviación
fig, ax = plt.subplots(figsize=(15,5))
# Generar la función de densidad
x = np.linspace(mu - 4*sigma, mu + 4*sigma, 1000)
pdf = 1/(sigma * np.sqrt(2 * np.pi)) * np.exp(-(x - mu)**2 / (2 * sigma**2))
# Graficar la función de densidad
plt.plot(x, pdf, color='red', lw=2)
# Títulos
plt.title(f'Distribución normal, media={mu}, desviación={sigma}')
plt.xlabel('Variable X')
plt.ylabel('Densidad de probabilidad')

# Show the plot
plt.show;


In [ ]:
import numpy as np
import matplotlib.pyplot as plt
# Media y desviación
mus = [50, 70]  # medias
sigmas = [5, 10]  # desviaciones
colors = ['red', 'blue']  # colores
# Fifura
fig, ax = plt.subplots(figsize=(15,5))
for mu, sigma, color in zip(mus, sigmas, colors):
    x = np.linspace(mu - 4*sigma, mu + 4*sigma, 1000)
    y = 1/(sigma * np.sqrt(2 * np.pi)) * np.exp(-(x - mu)**2 / (2 * sigma**2))
    ax.plot(x, y, color=color, label=f'media={mu}, desviación={sigma}')
# Título
ax.set_title('Dos distribuciones normales')
ax.legend()
# Gráfica
plt.show()


In [ ]:
#Normal inverse
from scipy.stats import norm
media=0
desviación=1
p = 0.975  # probabilidad
x = norm.ppf(p,media,desviación)  # inversa de la distribución normal en p
print(x)


In [ ]:
# Probabilidad
from scipy.stats import norm
x = 25  # Valor para encontar el percentil
media=25
desviación=2
p = norm.cdf(x,media,desviación)  # percentil de X en una distribución normal
print(p)

#***Seleccionar una muestra MAS***

In [ ]:
# Seleccionar una muestra aleatoria
df[['Country','GDPpp2020']].sample(n=3,random_state=1)

#***Prueba de diferencia de medias***

In [ ]:
df['Continent'].value_counts()

Africa       62
Europa       60
Americano    54
Asia         41
Oceania      20
Name: Continent, dtype: int64

In [ ]:
gdpAfrica=df['GDPpp2020'][df['Continent']=='Africa']
gdpAmerica=df['GDPpp2020'][df['Continent']=='Americano']

In [ ]:
gdpAfrica=gdpAfrica.dropna()
gdpAmerica=gdpAmerica.dropna()

In [ ]:
gdpAfrica.mean(),gdpAfrica.std()

(10099.071760006691, 16666.9680645212)

In [ ]:
gdpAmerica.mean(),gdpAmerica.std()

(23402.688660113323, 18147.55679879797)

In [ ]:
import numpy as np
from scipy.stats import ttest_ind


# Perform an independent t-test with equal variances and different sample sizes

t_stat, p_val = ttest_ind(gdpAfrica, gdpAmerica, equal_var=False)

# print the result
print("t-statistic:", t_stat)
print("p-value:", p_val)

t-statistic: -3.686529384431515
p-value: 0.00041460330995797735


#***Intervalo de confianza para la media***

In [ ]:
import numpy as np
import pandas as pd
from scipy import stats

# Caluclar la media muestral y la desviación estándar
sample_mean = np.mean(gdpAfrica)
sample_std = np.std(gdpAfrica, ddof=1)

# Fijar el nivel de confianza y calcular los grados de libertad
conf_level = 0.90
df = len(gdpAfrica) - 1

# Calcular el estadístico de prueba t y el margen de error
t_value = stats.t.ppf((1 + conf_level) / 2, df)
margin_of_error = t_value * (sample_std / np.sqrt(len(gdpAfrica)))

# Calcular el intervalo de confianza
limite_inferior = sample_mean - margin_of_error
limite_superior = sample_mean + margin_of_error

# Resultados
result = pd.Series([limite_inferior, limite_superior], index=['limite_inferior', 'limite_superior'])
print(f"El intervalo de confianza del {conf_level*100}% para la media poblacional es :\n{result}")



El intervalo de confianza del 90.0% para la media poblacional es :
limite_inferior     6439.869773
limite_superior    13758.273747
dtype: float64


#***Intervalo de confianza para la proporción***

In [ ]:
income=df['IncomeGroup']
n=len(income)
income
successes=income[income=='Low income'].count()
successes,n

(29, 237)

In [ ]:
from statsmodels.stats.proportion import proportions_ztest

# Definir la hipotesis nula de la proporción
null_prop = 0.05

# Calular la proporción muestral
sample_prop = successes / n

# Correr la prueba de hipótesis
stat, pval = proportions_ztest(successes, n, null_prop)

# REsultados del estadístico de prueba y el valor p
print(f"Z-score: {stat:.2f}")
print(f"P-value: {pval:.4f}")

Z-score: 3.40
P-value: 0.0007


In [ ]:
-Ayuda
Tabular una variable cualitativa X:
df['X’].value_counts(normalize=False)

Seleccionar solo las variables X, Y:
df[['X', 'Y']]

Filtrar cuando la variable X es x:
df[df['X']=='x']

Filtrar cuando la variable X=x & Y=y
df[(df['X']=='x') & (df['Y']='y') ]

Filtrar por texto be variable:
df[df['X'].str.contains('xxx')]

Ordenar por la variable X:
df.sort_values(by=['X'],ascending=True)

Bloquear por una o más condiones:
df.loc[(df['X'] == 'x') & (df['Y'] == 'y')]

Redondear decimales:
round(X,1)

Group by
df.groupby(['X','Y'])['Z'].aggregate(['count','mean','std’])

Crosstabulation
pd.crosstab(df['X'], df['Y'],normalize=False)

Crear una tabla dinámica
df.pivot_table(index='X',columns='Y',values='Z',aggfunc='mean')

